openhands/server/session/session.py

```diff
@@ -52,41 +53,31 @@ def __init__(
         self.agent_session.event_stream.subscribe(
             EventStreamSubscriber.SERVER, self.on_event, self.sid
         )
-        self.config = config
+        # Copying this means that when we update variables they are not applied to the shared global configuration!
+        self.config = deepcopy(config)
```

Comment:
This means we need one AppConfig per sid, with its other stuff. For other clients, when we need a specific instance of a config class, the config module is able to make them, under predefined names (like `draft_llm_config`) or user-defined names (like [these](https://github.com/All-Hands-AI/OpenHands/blob/c3ddb26e4302a8f41510e8c060a2c426263625f7/config.template.toml#L175)).

Maybe we can take that into account for these config instances? I don't think it needs to be in this PR.

Reply:
The reason I added this in here was the idea of writing back into the global config from the web input is really scary!
For example this line (I preserved the existing logic):
```default_llm_config.api_key = session_init_data.llm_api_key or default_llm_config.api_key```

Before the deep copy, this would write a user's api_key back into the global config. Another user could come along and call init without an API key (They would need to modify the client side code) and boom - they use an API key owned by the last person who entered one! (This was actually the primary motivation behind this PR.)

Reply:
This had me 😱 when I spotted it!

Reply:
Argh. By the way, this is why [these methods](https://github.com/All-Hands-AI/OpenHands/blob/de81020a8dc9ce1dfdcfddf7cf85fe0615b25c73/openhands/core/config/llm_config.py#L81) exist. Their only reason for existence is to be used by the UI to get defaults - actual defaults, dataclass values only, not from config, not from env, not from, uh, previous values.

Currently, the UI isn't using them, so nothing is using them. They're just sitting there and be pretty...